File: leetcode/product_of_array_except_self.cpp
#include <iostream>
#include <vector>

using namespace std;


class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        int num_zeros = 0;
        long non_zero_prod = 1;
        for (int &x : nums) {
            if (x != 0) {
                non_zero_prod *= x;
            } else {
                num_zeros++;
                if (num_zeros > 1) break;
            }
        }

        for (int &x : nums) {
            if (num_zeros > 1) {
                x = 0;
            } else if (num_zeros == 1) {
                x = (x == 0) * non_zero_prod;
            } else {
                x = (non_zero_prod / x);
            }
        }

        return nums;
    }
};


int main() {
    vector<int> nums = {1, 2, 0, 3, 4};

    Solution s;
    for (int &x : s.productExceptSelf(nums)) {
        cout << x << ' ';
    }
    cout << endl;

    return 0;
}

File: leetcode/neetcode_roadmap/product_of_array_except_self.py
class Solution:
    def productExceptSelf(self, nums: list[int]) -> list[int]:
        prod = 1
        num_zeros = 0
        for num in nums:
            if num == 0:
                num_zeros += 1

                if num_zeros > 1:
                    return [0 for _ in nums]
            else:
                prod *= num

        if num_zeros == 1:
            return [0 if n != 0 else prod for n in nums]

        return [int(prod/n) if n != 0 else 0 for n in nums]


S = Solution()

nums = [-1,1,0,-3,3]
print(S.productExceptSelf(nums))
